eventbuilding with data taken with new DAQ

kunathj opened the pull request:
Hi @airqui
As discussed, I tried to rewrite the eventbuilding code (`_converted.root` → `_build.root`) so that it works with the new setup (14 or 15 slabs with new DAQ).
I haven't done a full validation yet. I'm still missing the CERN eos logins, and only have the run `run_050014_07272021_16h35min_Ascii` locally. 
I don't know how to produce the masked/pedestals/mip_calib files myself.
Thus, I ran the above-mentioned 15-slab run with the PROTO14-calibration.
This at least allowed me to check that the eventbuilding can run without an error. 

```sh
export COMMISSIONING_TAG="_PROTO14_run_050010"
export RUN="run_050014_07272021_16h35min_Ascii"
export data_slb_converted="/home/kunath/SiWECAL-TestBeam/converter_SLB/convertedfiles/${RUN}/${RUN}_converted.root"
export N_EVENTS=300
cd eventbuilding
  ./build_events.py $data_slb_converted $COMMISSIONING_TAG -n $N_EVENTS -w 0
```

Before I continue, I have a couple of questions:
1. Can you guide me through the creation of the masked/pedestals/mip_calib files for _my_ run?
2. There are a few places in the code marked with *TODO*. There, input would be helpful.
3. Here I am overwriting the old eventbuilding files. Is this correct, or do we need to keep them in this branch?
4.  If overwriting is ok, I am assuming it would also be a good idea to remove all those files that are now useless (e.g. all the old pedestal files, etc.

If you prefer to discuss this over zoom, that is also fine.

Cheers

Jonas

kunathj committed:
Make python script python2/3 compatible

kunathj committed:
Adapt the eventbuilding CLI

kunathj committed:
Fix errors: remove tdc and rename input tree

With the new SLBoard data reading setup, TDC is never used.
The incoming root file's tree (built in converter_SLB)
is currently named "siwecaldecoded" instead of the old "ecal".

kunathj committed:
eventbuilding: Merge config reading into one class

Before, the reading of the configuration options was done
through sperate functions and with global variables.
This is now merged into a single EcalConfig class.

This commit should only have one functionality change:
It is now possible to specify the parent directory of all config files.
Some apparent bugs and improvement opportunities ere identified.
These are NOT change in this commit, in order to not hide
their change under the large-scale refactoring that is performed here.

kunathj committed:
Combine the global constants into a class

kunathj committed:
Fix the W configuration

Fix the multiplication-bug.
By moving the W config into the EcalNumbers class,
it can now be changed during runs and without
changing the python file itself.

kunathj committed:
Eventbuilding: change build_events into a class

kunathj committed:
Eventbuilding: Steer commissioning dir from CLI.

kunathj committed:
eventbuilding: Put BCID merging into a class.

kunathj committed:
Clean up the branch filling

kunathj committed:
Clean up the EcalHit class

kunathj committed:
Add a new flag to the build file: isCommissioned

kunathj committed:
Adapt to SLB_commissining: All in one file now

airqui commented:
> Hi @airqui As discussed, I tried to rewrite the eventbuilding code (`_converted.root` → `_build.root`) so that it works with the new setup (14 or 15 slabs with new DAQ). I haven't done a full validation yet. I'm still missing the CERN eos logins, and only have the run `run_050014_07272021_16h35min_Ascii` locally. I don't know how to produce the masked/pedestals/mip_calib files myself. Thus, I ran the above-mentioned 15-slab run with the PROTO14-calibration. This at least allowed me to check that the eventbuilding can run without an error.
> 
> ```shell
> export COMMISSIONING_TAG="_PROTO14_run_050010"
> export RUN="run_050014_07272021_16h35min_Ascii"
> export data_slb_converted="/home/kunath/SiWECAL-TestBeam/converter_SLB/convertedfiles/${RUN}/${RUN}_converted.root"
> export N_EVENTS=300
> cd eventbuilding
>   ./build_events.py $data_slb_converted $COMMISSIONING_TAG -n $N_EVENTS -w 0
> ```
> 
> Before I continue, I have a couple of questions:
> 
>     1. Can you guide me through the creation of the masked/pedestals/mip_calib files for _my_ run?
> 
>     2. There are a few places in the code marked with _TODO_. There, input would be helpful.
> 
>     3. Here I am overwriting the old eventbuilding files. Is this correct, or do we need to keep them in this branch?
> 
>     4. If overwriting is ok, I am assuming it would also be a good idea to remove all those files that are now useless (e.g. all the old pedestal files, etc.
> 
> 
> If you prefer to discuss this over zoom, that is also fine.
> 
> Cheers
> 
> Jonas

HI Jonas,
Thanks for the pull request and sorry for the slow reaction.

-I would not use the run14: it is too short and you won't be able to extract enough information for pedestal and mip calibration.
The run 10 is better suited although it has 14 slabs.
Today and tomorrow I am finishing the commissioning for cosmics. Monday I will start a run of cosmics. By Wednesday you would be able to already start looking at the new data (which are the ones to be trusted more).

- For the pedestal and mip, you should go to the SLBperformance folder and use the 
analysis_run_050xxx_pedestal.sh (or mip).
However, I already tell you that this is a very tedious process because we are dealing with too many files and too many cells.
For a run, the DAQ divides the data taking in files with the sufix .dat_XXXX

Pedestal:
What I do is for every single .dat_XXXX I create a set of histograms with pedestal information.
Once I have the root files for all the data files, I merge them and I calculate the pedestals 15slabs x 16 chips x 64 channels x 15 memory cells (for cosmics only 0-2 or so memory cell are usable)

This should produce a pedestal file (maybe with different naming, to be checked)

MIP:
Similar process but using the previously calculated pedestal values.
For every file I create the  signal histograms (ADC-pedestal), I merge them when I have all of them processed and then I calculate the MIPS.

This process should produce the MIP file.
about 3/4: yes we should overwrite or remove accordingly (the mergeroot files are not needed anymore).

About your TODOs:

# TODO: I understand why bcid = -999 is the filler value from converter_SLB. But why is bcid < 50 not ok?"
This is a pure technical issue: there is more noise at the begining of the acquisition (when the start acquisiton , power up of the chips etc)

# TODO: Should corrected_bcid be used instead of bcid?"
In principle yes. The corrected bid accounts for the overrunning of the 12 bits. However we expect to have no overruning in the coming testbeam (to be confirmed). Therefore, corrected_bcid and bcid should be the same.

# TODO What are the real pos_z and Tungsten values for 15 slabs?"
The slab position is stored in the ntuple in the "slot" variable. (this may have changed since our conversation few weeks ago)
Slot 0 means first slab, in the position 0.
Slot 1, in position 1 etc.. The distance between positions is 15mm. (position 0 should be z=0)

# TODO: Is this really doing the right thing to the positioning? Check with new cosmic runs."
The issue with the mapping was not real. The data that you were using was simply not okay. 
The x-y mapping is the same for all slabs. Still possible misalignments are possible (and to be corrected offline)

# TODO: Is it really ok to just us the pedestal mean from the first SCA?
No!! Every SCA has its own pedestal and it should be used.
If a SCA has no pedestal value we have two options: tag this SCA as "notCalibrated" or use the average pedestal of the others.
In principle, if in the beam test we have many of these cases, we should investigate further the reason.

 # TODO: Are we really ok with 0-entries in the pedestal map?
Not sure what you mean here.

Cheers,
Adrián

airqui commented:
@kunathj 

I forgot about the masked channels.
For every run, we store the settings used in a file called Run_Settings.txt
There is a simple macro in SLBcommissioning that reads this file and creates the list of masked channels.

test_read_masked_channels_summary.C

kunathj commented:
Ok, I will wait for that new cosmics run (and main CERN logins...).

1. So I can just dynamically read out the `pos_z` from the `_converted`-file, as `pos_z = 15 * slab`? That is cool, one less steering parameter that is needed!
2. I would argue that the `x0` (↔ amount of Tungsten in front of the cell) should be dropped from the build 
    - If you know the W-config of the run, it is very easy to later calculate it from `pos_z`.
    - When building the build-file, you would have to remember each time to change the W-config in the code to the one used in the run. This seems like a mistake waiting to happen.
    - Technicality, but: The actual value is only true for a particle going orthogonally through the detector. 
    - I could be convinced that it is good to store the W-config information in the same file. Even then, I would rather store it in a seperate tree.
  
## 3. pedestal

The previous code used only the map value from SCA[0] and filled all scas of the channel with that value.
It also had a check `if ped_norm <=5: self.isMasked = 1`.  This check of course was never triggered, as all 15 scas were the same.

I would do the following: 
- If the sca has a pedestal value, use that one. 
- If not, but there is at least 5 other pedestals that have one: use the average and tag the sca as `not isCommissioned`.
- If the sca does not have a pedestal value, and there is not enough other scas with a pedestal value in the chip: No pedestal substraction and flag tag this sca as `isMasked`.

Does this sound reasonable?

airqui commented:
Hi,

"So I can just dynamically read out the pos_z from the _converted-file, as pos_z = 15 * slab?"
I agree that this is nice, but I prefer not doing it like that, sorry. The reason is that the hardware is a living thing (one slab may die or removed or position swapped...) We should always keep flexibiility. The default option should be  pos_z = 15 * slab, but others are also possible.

Similar about the tungsten. The tungsten repartition may change during the time and the optimal way is to keep this information IN the file. I agree though, that we don't need it for every event. Probably a simple TH1F or whatever is okay for you (we only need to know only the amount of tungnsten / X0 in front each layer).

Your pedestal related proposition seems reasonable. The use of only the SCA[0] was a mistake or probably a change from someone doing tests and not reverting them back.. (someone being probably me, I guess)

kunathj commented:
> "So I can just dynamically read out the pos_z from the _converted-file, as pos_z = 15 * slab?"

I meant to write `pos_z = 15 * slot`. As you point, out, slabs can of course be moved around. But the way I understand `slot` from the code base, `pos_z = 15 * slot` should always be true (until you change the mechanical structure, or use a _long slab_). But if you still prefer it, I can of course leave the CLI-steering option for it in the code.

Yes, I will go with the TH1F for the Tungsten.

I'll push the updated version as soon as I get access to the eos, to validate the procedure.

Thanks for your repeated help @airqui

airqui commented:
pos_z = 15 * slot is always right... in principle.
The thing is that the "slot" is not something that the DAQ knows automatically: the slot is hardcoded in the SLboard with a physical  swtich. This means that we can swap slab positions but the slot-id written in the data won't be changed.

In any case. After some thinking, let's do as you suggest. Few months ago, not everyone playing with the slabs knew how to do this but this is not the case anymore. If there is any swap of positions, the slot-id will also changed accordingly.

@kunathj

kunathj committed:
Add a Makefile to steer the event building

kunathj committed:
Use corrected bcid (should be the same in cosmics)

kunathj committed:
Guide the building through a Makefile

kunathj committed:
Makefile: Create no files in the ascii data folder

It is possible that we do not have write access on the data folder.

kunathj committed:
Add changes discussed in PR

[eventbuilding with new DAQ](https://github.com/SiWECAL-TestBeam/SiWECAL-TB-analysis/pull/17)

kunathj committed:
Remove deprecated files

kunathj committed:
Update documentation

kunathj commented:
Hi @airqui,

I think this is finally ready to be merged. 🥳
I ran the whole reconstruction chain with the latest commit, and arrive at the `_build.root` file for the `run_050016` _(I am a bit cheating with this statement. As mentioned earlier, the final eventbuilding step is rather slow. I only ran this step on a subset of the data so far; the full statistics are running currently)_.

Below you can find a picture of one of the muon candidates.
![2D_slices_4](https://user-images.githubusercontent.com/37667761/139618282-da0a2b63-2809-4932-bdfb-d65c0bbbb012.png)

kunathj committed:
Typo fixes

kunathj committed:
Improve new isCommissioned flag

- If the used SCA is calibrated, we do not care about the others.
- Requiring 5 other SCA pedestals to be calibrated is too much
  (at least in a cosmics run). Reduced to 3.

airqui commented:
Hi @kunathj 

Thanks for the nice changes! You really went deep in the code. Thanks!

However, I have few requests.

- Makefile... I don't think that you should merge in one automatic thing the pedestal/mip/monitoring tools and the building stuff. One is commissioning/debugging level, the other is an analysis step.
In fact, we cannot do the proper event building before the 2-3 days of mip calibration . For the moment we will use dummy built events.
Please reconsider using the Makefile only for the buliding steps.
If you want, you can create an automatic "cosmic" analysis script or makefile or whatever. But not for default analysis. 
Cosmic and beam data are  quite dfferent.

- I see your changes in the folder SLBcommissioning, they look in principle good and in the right direction, but this is the folder that we use for the initial debugging of the prototype. Please, remove them from the pull request (we could merge them after the beam test). First rule of a data taking run: never change a running system!

Thanks!
Adrián

airqui commented:
> Hi @airqui,
> 
> I think this is finally ready to be merged. partying_face I ran the whole reconstruction chain with the latest commit, and arrive at the `_build.root` file for the `run_050016` _(I am a bit cheating with this statement. As mentioned earlier, the final eventbuilding step is rather slow. I only ran this step on a subset of the data so far; the full statistics are running currently)_.
> 
> Below you can find a picture of one of the muon candidates. ![2D_slices_4](https://user-images.githubusercontent.com/37667761/139618282-da0a2b63-2809-4932-bdfb-d65c0bbbb012.png)

Nice!

kunathj commented on README.md:
````diff
@@ -13,14 +14,36 @@ cd singleslab/
+  RAW_DATA_DIR=/data_ilc/flc/ECAL/cosmics/TB2021-11/run_050016_10192021_21h49min_Ascii
+```
+
+It is recommended to split the make run into parts:
+
+1. `make converted`
+2. `make masked`
+3. `make pedestals`
+4. `make mip_calib`
+5. `make run`
+
+Then, check the results of each step before proceding to the next.
````
@airqui So y want me to mention here explicitely that these 5 make steps are _only_ run like that for cosmic runs?
Or is that itself a wrong statement?

kunathj commented:
### SLBcommissioning

That is a very valid remark. I disentangled the changes to the SLBcommissioning folder.
    Those that would be nice to apply later on (for consistency with other changes that I made) are now tracked in a pull request to my fork: https://github.com/kunathj/SiWECAL-TB-analysis/pull/1
    I will keep them there for now, and we can consider including them after the TB, according to your suggestion.

### Makefile

My idea here was not to suggest an _automatic_ `make` (or `make build`) call. Of course each step should first be validated before going on to the next one. I tried to convey this idea through my remark in the README https://github.com/SiWECAL-TestBeam/SiWECAL-TB-analysis/pull/17#discussion_r741877556.

The advantage I see with this Makefile: There is one central place to change stuff. So e.g. if my data folder is different to the one used by the last person committing to the repo (read: your data is on the eos, mine somewhere else), then there is only this one line in the Makefile that I have to change, instead of going through all the commissioning and build scripts.

If you have a strong negative opinion on this Makefile, I could rename it (`cosmics.mk` or `Makefile-Jonas.mk`)?
One Makefile-advantage that I do not want to leave unmentioned: It allows some basic multiprocessing (`make -j` was a good friend to me the last few days).

airqui commented:
Hi, Please rename the makefile or add the info to make it clear that it is for a particular case (as I think that you mention in the thread).
For the conflicts, solve the gitignore one and "uncommit" all changes on SLBperformance for the time being.

Thanks!

kunathj committed:
Speed up the Makefile

kunathj committed:
Progress prints now overwrite previous progress

kunathj committed:
Make the cob positions a function argument

kunathj committed:
Debugging the script with a long cosmics run

kunathj committed:
Rename the cosmics Makefile

kunathj committed:
Merge branch 'slboard_TB2020' into slboard_TB2020

kunathj commented:
Ok, in the end I was definitely trying to do too many things at once here.
I decluttered this, and will potentially open dedicated PRs for other changes.
Now, these changes do not touch any existing code (they add one default argument to `converter_SLB/ConvertDataSL.cc`, but that's it).

This means that you should be able to merge this without any concerns.
At the same time, this also means that fixes that I had to apply for the cosmics building are _not_ here.
E.g., with this current code most MIP fits will fail.
But as this does not really have anything to do with the build script, it should be the topic of its own PR.
